# Queueing theory - M/M/1 example

### 1. Imports

In [1]:
import openqtsim

### 2. Prepare inputs

In [2]:
# Basic input
t_scale = 1 # 1: hours, 60: minutes, 3600: seconds, etc
Lambda = 8
Mu = 9

In [3]:
# Create Arrival and Service processes and specify the number of servers
A = openqtsim.ArrivalProcess("M", arr_rate=Lambda, t_scale=t_scale)
S = openqtsim.ServiceProcess("M", srv_rate=Mu, t_scale=t_scale)
c = 1

In [4]:
# Create the queue object and use it to create a simulating object
q = openqtsim.Queue(A, S, c)
sim = openqtsim.Simulation(q, t_scale=t_scale, seed=1)

### 3. Run the simulation

In [5]:
# set stop limits
tol = 0.0001
max_arr = 50000

In [6]:
%%time
print('Start simulation of "{}" queue'.format(q.kendall_notation))
print('')

# run simulation - simulation stops either when max arrivals is reached or the tolerance limits are achieved
sim.run(max_arr=max_arr, IAT_tol=tol, ST_tol=tol) 

Start simulation of "M/M/1/inf/inf/FIFO" queue

Nr of customers: 10777
Accuracy IAT: 9.519399568647113e-05
Tolerance: 0.0001
Accuracy ST: 9.643113860524632e-05
Tolerance: 0.0001

Wall time: 15.2 s


### 4. Analyse and visualise results

#### Get statistics

In [7]:
print('--- input statistics --------')
print("aver_IAT_in_secs: {:.4f} [hour]".format(1/Lambda))
print("aver_ST_in_secs: {:.4f} [hour]".format(1/Mu))
print('')

print('--- simulation statistics ---')
sim.get_stats()

--- input statistics --------
aver_IAT_in_secs: 0.1250 [hour]
aver_ST_in_secs: 0.1111 [hour]

--- simulation statistics ---
Average IAT: 0.1249 [hours]
Average ST: 0.1112 [hours]
Rho: System utilisation: 0.8902

Total number of customers: 10777.00

W_s: Average time spent in the system: 0.85 [hours]
W_q: Average time spent in the queue: 0.74 [hours]
Average nr arrivals: 8.00 [# per hour]
Average nr services: 8.99 [# per hour]

Total waiting time: 7957.1539 [hours]
Average waiting time of all customers: 0.74 [hours]
Average waiting time of customers that waited: 0.83 [hours]

P_0: Probability of idle server (nobody in the system): 0.1098
Probability that somebody is waiting: 0.8917
Probability that nobody is waiting: 0.1083
P_2: Probability that 1 person is waiting: 0.1051
P_3: Probability that 2 persons are waiting: 0.0886
P_4: Probability that 3 persons are waiting: 0.0766
P_10: Probability that 9 persons are waiting (10 in system): 0.0423

Average waiting time as a fraction of ST: 6.

**Analytical solution: M_M_1_inf_inf_FIFO**

    arrival rate: 8.0000 per hour
    service rate: 9.0000 per hour
    rho: 0.8889

    P_0: probability that there are 0 customers in the system = 0.1111
    P_1: probability of 1 customers in the system = 0.0988
    P_10: probability of 10 customers in the system = 0.0342

    no_q: probability that there is nobody in the queue (P_0+P_1) = 0.2099
    util_sys: system utilisation rate  (P_(n>=1)) = 0.8889
    util_ser: server utilisation rate (P_(n>=2)) = 0.7901

    L_s: the long term average number of customers present in the system = 8.0000
    L_q: the long term average number of customers present in the queue = 7.1111
    W_s: the long term average time spent in the system = 1.0000 hour (= 60.0 minutes)
    W_q: the long term average waiting time in the queue = 0.8889 hour (= 53.3 minutes)

#### Return log

In [8]:
sim.return_log()
# c = customer
# IAT = inter arrival time
# ST = service time
# AT = arrival time
# TSB = time service begins
# TSE = time service ends
# TCSS = time customer spends in the system
# TCWQ = time customer waits in the queue
# ITS = idle time of the server
# QL = queue length

c_id       IAT        ST           AT          TSB          TSE  \
0          1  0.067451  0.000013     0.067451     0.067451     0.067463   
1          2  0.159266  0.017634     0.226717     0.226717     0.244351   
2          3  0.045002  0.022902     0.271718     0.271718     0.294620   
3          4  0.012110  0.056161     0.283828     0.294620     0.350781   
4          5  0.052997  0.060371     0.336826     0.350781     0.411152   
5          6  0.096745  0.025414     0.433571     0.433571     0.458984   
6          7  0.144485  0.003086     0.578056     0.578056     0.581141   
7          8  0.263087  0.060010     0.841143     0.841143     0.901153   
8          9  0.138760  0.016808     0.979903     0.979903     0.996711   
9         10  0.102251  0.179241     1.082154     1.082154     1.261395   
10        11  0.027597  0.041782     1.109751     1.261395     1.303177   
11        12  0.431278  0.232291     1.541029     1.541029     1.773320   
12        13  0.147338  0.009876     1.688367     1.773320     1.783195   
13        14  0.281257  0.020681     1.969624     1.969624     1.990304   
14        15  0.004980  0.011503     1.974604     1.990304     2.001807   
15        16  0.263113  0.351940     2.237716     2.237716     2.589656   
16        17  0.068330  0.128884     2.306046     2.589656     2.718541   
17        18  0.095223  0.002051     2.401269     2.718541     2.720592   
18        19  0.147157  0.154097     2.548426     2.720592     2.874688   
19        20  0.047386  0.153220     2.595812     2.874688     3.027909   
20        21  0.224943  0.036569     2.820755     3.027909     3.064478   
21        22  0.562164  0.012106     3.382919     3.382919     3.395025   
22        23  0.194653  0.265829     3.577572     3.577572     3.843400   
23        24  0.074252  0.026421     3.651824     3.843400     3.869821   
24        25  0.043449  0.034292     3.695273     3.869821     3.904113   
25        26  0.042420  0.006093     3.737693     3.904113     3.910206   
26        27  0.017412  0.094844     3.755105     3.910206     4.005050   
27        28  0.002445  0.133685     3.757549     4.005050     4.138735   
28        29  0.141975  0.131720     3.899524     4.138735     4.270455   
29        30  0.084554  0.121115     3.984079     4.270455     4.391570   
...      ...       ...       ...          ...          ...          ...   
10747  10748  0.016576  0.119680  1341.566438  1342.229670  1342.349350   
10748  10749  0.326050  0.118306  1341.892488  1342.349350  1342.467656   
10749  10750  0.477527  0.017280  1342.370015  1342.467656  1342.484936   
10750  10751  0.010796  0.292445  1342.380811  1342.484936  1342.777381   
10751  10752  0.240203  0.030820  1342.621013  1342.777381  1342.808201   
10752  10753  0.104800  0.141822  1342.725813  1342.808201  1342.950023   
10753  10754  0.583419  0.102228  1343.309232  1343.309232  1343.411459   
10754  10755  0.052048  0.366651  1343.361280  1343.411459  1343.778110   
10755  10756  0.136368  0.203339  1343.497648  1343.778110  1343.981449   
10756  10757  0.111117  0.124270  1343.608766  1343.981449  1344.105720   
10757  10758  0.125046  0.057178  1343.733811  1344.105720  1344.162897   
10758  10759  0.041206  0.145957  1343.775017  1344.162897  1344.308854   
10759  10760  0.205287  0.055090  1343.980304  1344.308854  1344.363944   
10760  10761  0.064248  0.012737  1344.044552  1344.363944  1344.376681   
10761  10762  0.045557  0.045945  1344.090110  1344.376681  1344.422625   
10762  10763  0.086471  0.081359  1344.176580  1344.422625  1344.503984   
10763  10764  0.072522  0.167017  1344.249102  1344.503984  1344.671001   
10764  10765  0.024166  0.032474  1344.273268  1344.671001  1344.703475   
10765  10766  0.126551  0.021442  1344.399820  1344.703475  1344.724916   
10766  10767  0.125687  0.011107  1344.525507  1344.724916  1344.736023   
10767  10768  0.433076  0.049632  1344.958583  1344.958583  1345.008215   
10768  10769  0.120278  0.020976 

In [9]:
import matplotlib.pyplot as plt
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 12))

fig.suptitle('Overview of stochastic variables')
ax1.hist(sim.log["IAT"],1000); ax1.set_title('IAT');
ax2.hist(sim.log["ST"],1000); ax2.set_title('ST');

## Sandbox

### Small test M//M/1 queue simulator

In [161]:
from scipy import stats
import numpy as np
import pandas as pd

In [162]:
def get_IAT_and_ST(lam, mu, nr_arr, seed):# generate list of inter arrival times
    
    rv_iat = stats.expon(scale = 1/lam)
    rv_st = stats.expon(scale = 1/mu)
    
    # generate list of inter arrival times
    np.random.seed(seed)
    IAT = rv_iat.rvs(nr_arr)

    # generate list of service times
    np.random.seed(seed)
    ST = rv_st.rvs(nr_arr)
    
    return IAT, ST

In [163]:
def calculate_MM1(IAT, ST):
    
    df = pd.DataFrame()
    AT = []
    TSB = []
    TSE = []
    TCSS = []
    TCWQ = []
    ITS = []
    for i in range(len(IAT)):
        if i == 0:
            AT.append(IAT[i])
        else:
            AT.append(AT[i-1]+IAT[i])
        if i == 0:
            TSB.append(AT[i])
        else:
            TSB.append(np.max([AT[i], TSE[i-1]]))
        TSE.append(TSB[i]+ST[i])   # moment service begins plus service time
        TCSS.append(TSE[i]-AT[i])  # moment of arrival until service ends
        TCWQ.append(TSB[i]-AT[i])  # moment of arrival until service begins
        if i == 0:
            ITS.append(IAT[i])  # the server will start idle until the first arrival
        else:
            ITS.append(np.max([AT[i]-TSE[i-1], 0]))  # todo: I don't think this works with multiple servers

    df["IAT"]=IAT
    df["ST"]=ST
    df["AT"]=AT
    df["TSB"]=TSB
    df["TSE"]=TSE
    df["TCSS"]=TCSS
    df["TCWQ"]=TCWQ
    df["ITS"]=ITS
    
    return df

In [197]:
def get_stats(df):
    
    value = np.mean(df["TCWQ"])/np.mean(df["ST"])
    print('Waiting time over service time: {:.2f}'.format(value))

    value = np.mean(df[df["TCWQ"]!=0])/np.mean(df["ST"])
    print('Waiting time over service time (for customers that waited): {:.2f}'.format(value))
    
    value = (df["TSE"].iloc[-1] - np.sum(df["ITS"]))/df["TSE"].iloc[-1]
    print('Rho: system utilisation: {:.4f}'.format(value))
    
    value = np.sum(df["ITS"]) / df["TSE"].iloc[-1]
    print('P_0: probability nobody in the system: {:.4f}'.format(value))


In [199]:
value = np.mean(df[df["TCWQ"]!=0])/np.mean(df["ST"])
value

IAT         0.856740
ST          0.761546
AT      45046.110727
TSB     45048.268906
TSE     45049.030453
TCSS        2.919726
TCWQ        2.158180
ITS         0.000000
dtype: float64

In [198]:
IAT, ST = get_IAT_and_ST(8, 9, 80000, 5)
df = calculate_MM1(IAT, ST)
get_stats(df)

Waiting time over service time: 1.92


TypeError: unsupported format string passed to Series.__format__